<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#-Homework-1-(Topics-1-4:-Data-Structures,-Indexing,-Cleaning-Data,-Math)" data-toc-modified-id="-Homework-1-(Topics-1-4:-Data-Structures,-Indexing,-Cleaning-Data,-Math)-1"><span class="toc-item-num">1&nbsp;&nbsp;</span><center> Homework 1 (Topics 1-4: Data Structures, Indexing, Cleaning Data, Math)</center></a></span><ul class="toc-item"><li><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Task-1." data-toc-modified-id="Task-1.-1.0.0.1"><span class="toc-item-num">1.0.0.1&nbsp;&nbsp;</span>Task 1.</a></span></li><li><span><a href="#Варианты-ответов:" data-toc-modified-id="Варианты-ответов:-1.0.0.2"><span class="toc-item-num">1.0.0.2&nbsp;&nbsp;</span>Варианты ответов:</a></span></li><li><span><a href="#Task-2." data-toc-modified-id="Task-2.-1.0.0.3"><span class="toc-item-num">1.0.0.3&nbsp;&nbsp;</span>Task 2.</a></span></li><li><span><a href="#Вопросы:" data-toc-modified-id="Вопросы:-1.0.0.4"><span class="toc-item-num">1.0.0.4&nbsp;&nbsp;</span>Вопросы:</a></span></li><li><span><a href="#Task-3." data-toc-modified-id="Task-3.-1.0.0.5"><span class="toc-item-num">1.0.0.5&nbsp;&nbsp;</span>Task 3.</a></span></li><li><span><a href="#Выберите-все-верные-пункты:" data-toc-modified-id="Выберите-все-верные-пункты:-1.0.0.6"><span class="toc-item-num">1.0.0.6&nbsp;&nbsp;</span>Выберите все верные пункты:</a></span></li><li><span><a href="#Task-4." data-toc-modified-id="Task-4.-1.0.0.7"><span class="toc-item-num">1.0.0.7&nbsp;&nbsp;</span>Task 4.</a></span></li><li><span><a href="#Пояснения:" data-toc-modified-id="Пояснения:-1.0.0.8"><span class="toc-item-num">1.0.0.8&nbsp;&nbsp;</span>Пояснения:</a></span></li></ul></li><li><span><a href="#Информация-для-последующих-заданий" data-toc-modified-id="Информация-для-последующих-заданий-1.0.1"><span class="toc-item-num">1.0.1&nbsp;&nbsp;</span>Информация для последующих заданий</a></span></li></ul></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Описание данных</a></span><ul class="toc-item"><li><span><a href="#Таблица-transactions.csv" data-toc-modified-id="Таблица-transactions.csv-1.1.1"><span class="toc-item-num">1.1.1&nbsp;&nbsp;</span>Таблица <code>transactions.csv</code></a></span></li><li><span><a href="#Таблица-gender_train.csv" data-toc-modified-id="Таблица-gender_train.csv-1.1.2"><span class="toc-item-num">1.1.2&nbsp;&nbsp;</span>Таблица <code>gender_train.csv</code></a></span></li><li><span><a href="#Таблица-tr_mcc_codes.csv" data-toc-modified-id="Таблица-tr_mcc_codes.csv-1.1.3"><span class="toc-item-num">1.1.3&nbsp;&nbsp;</span>Таблица <code>tr_mcc_codes.csv</code></a></span></li><li><span><a href="#Таблица-tr_types.csv" data-toc-modified-id="Таблица-tr_types.csv-1.1.4"><span class="toc-item-num">1.1.4&nbsp;&nbsp;</span>Таблица <code>tr_types.csv</code></a></span><ul class="toc-item"><li><span><a href="#Task-5." data-toc-modified-id="Task-5.-1.1.4.1"><span class="toc-item-num">1.1.4.1&nbsp;&nbsp;</span>Task 5.</a></span></li><li><span><a href="#Task-6." data-toc-modified-id="Task-6.-1.1.4.2"><span class="toc-item-num">1.1.4.2&nbsp;&nbsp;</span>Task 6.</a></span></li><li><span><a href="#Выберите-все-верные-пункты:" data-toc-modified-id="Выберите-все-верные-пункты:-1.1.4.3"><span class="toc-item-num">1.1.4.3&nbsp;&nbsp;</span>Выберите все верные пункты:</a></span></li><li><span><a href="#Task-7." data-toc-modified-id="Task-7.-1.1.4.4"><span class="toc-item-num">1.1.4.4&nbsp;&nbsp;</span>Task 7.</a></span></li><li><span><a href="#Выберите-все-верные-пункты:" data-toc-modified-id="Выберите-все-верные-пункты:-1.1.4.5"><span class="toc-item-num">1.1.4.5&nbsp;&nbsp;</span>Выберите все верные пункты:</a></span></li><li><span><a href="#Task-8." data-toc-modified-id="Task-8.-1.1.4.6"><span class="toc-item-num">1.1.4.6&nbsp;&nbsp;</span>Task 8.</a></span></li></ul></li></ul></li></ul></li></ul></div>

# <center> Homework 1 (Topics 1-4: Data Structures, Indexing, Cleaning Data, Math)

In [17]:
import pandas as pd
import numpy as np

#### Task 1. 
Задание:

1. Создайте Series из последовательности 15 значений, равномерно разбивающих отрезок [0, 20] (воспользуйтесь функцией linspace)
2. Определите отношение элементов полученной серии к их предыдущим элементам (\*). 
3. В результате необходимо получить среднее полученного вектора, оставив в нём только те значения, которые не более чем 1.5 (\**). 

Выберите из ответов тот, который максимально близок к полученному (c точки зрения абсолютной разницы).

#### Варианты ответов:
    - 1) 1.24
    - 2) 1.18 
    - 3) 0.71
    - 4) 1.13


##### Пояснения:
(\*) Если было бы необходимо найти последовательность из 3-х значений,  
равномерно разбивающих отрезок [0,1], то это были бы значения [0, 0.5, 1]. 

(\**) Если был бы дан список элементов a = [1,2,3,12], отношения элементов к предыдущим будут равны [NaN, 2, 1.5, 4].  
А на последнем этапе в таком примере останется только [1.5] и среднее значение будет также 1.5.

In [18]:
n1 = pd.Series(data = np.linspace(0, 20, 15))
n2 = n1.divide(n1.shift(1))
n3 = n2[n2 <= 1.5].mean()
n3

1.181677812927813

Получили значение равное 1.181677812927813
Ответ: 2

#### Task 2. 
Выберите все верные ответы касательно следующих 3-х Series:
    - pd.Series('abcde'); (1) 
    - pd.Series(['abcde']); (2) 
    - pd.Series(list('abcde')); (3)
    - pd.Series("abcde"); (4)
    
##### Пояснения:
(\*)
- функция list: в строке каждый символ - это отдельный элемент для list
- квадратные скобки: в квадратных скобках списку передается множество элементов по отдельности через запятую


#### Вопросы:
    - 1) Серия (1) совпадает с серией (2), так как в каждом из случаев серия создаётся из списка строк
    - 2) Серия (2) совпадает с серией (3), так как в каждом из случаев серия создаётся из списка символов 
    - 3) Серия (1) не совпадает с серией (4), так как в (4) используются двойные кавычки "" вместо одинарных '' 

Ответы: 1, 3

#### Task 3. 

По клиенту получены зашумленные данные (объект s типа Series) по его транзакциям. 

Для заданного ниже объекта s проделайте следующее:
1. Создайте новый Series, значения которого совпадают со значениями s, а индексы - целочисленные значения от 2 до 12, не включая 12. 
2. Выберите из s элементы с индексами 3 и 5, после чего просуммируйте их, сохранив результат (1). 
3. Выберите из s только целочисленные элементы и вычислите их дисперсию (2). (\*)

Все полученные результаты округлите до 2-х знаков после запятой.

#### Выберите все верные пункты:
    - 1) Ответ (1) - 642.52
    - 2) Ответ (1) - 91.78
    - 3) Ответ (1) - не может быть определён
    - 4) Ответ (2) - 57591.19
    - 5) Ответ (2) - 210.12
    - 6) Ответ (2) - не может быть определён

##### Пояснения:
(\*) Целочисленные значения - значения, имеющие тип int.
- Дисперсия рассчитывается с помощью функции из библиотеки numpy: np.var( , ddof=0) или встроенной в python функции: .var(ddof=1)

In [19]:
s = pd.Series(data=['1', 2, 3.1, 'hi!', 5, -512, 12.42, 'my_name', 10.10, 98], 
              index=range(6, 26, 2))
s2 = pd.Series(data=list(s), index=range(2, 12))
s2

2           1
3           2
4         3.1
5         hi!
6           5
7        -512
8       12.42
9     my_name
10       10.1
11         98
dtype: object

In [20]:
s2.loc[3] + s2.loc[5]

TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [ ]:
s2[s2.apply(lambda x: isinstance(x, int))].var(ddof=0)

Пункт 2 не может быть выполнен
Дисперсия равна 57591.19
Ответы: 3, 4

#### Task 4. 
1. Сгенерируйте Series из 100 значений нормально распределённой СВ (np.random.normal с дефолтными параметрами - нулевым средним и единичной дисперсией).
2. Возведите каждое значение серии в 3 степень, а значения индекса увеличьте в 3 раза. 
3. Ответьте на следующие вопросы через запятую (без пробелов) (\*)
    - 1) Выведите сумму элементов, строго меньших 2.6, имеющих нечётные значения индекса. 
    - 2) Выведите количество значений серии меньше нуля.

#### Пояснения:
(\*) Если получились ответы 3, 4.32, то в форму необходимо внести их в виде "3,4.32". То есть вещественные числа необходимо разделять точками. Не забудьте про фиксированный seed (его менять не нужно)!
 - Определенное значение seed нужно, чтобы ответы у всех выполняющих
это задание были одинаковые и их можно было проверить (так как генерируются
одинаковые series).
 - Следует внимательнее использовать [ ] для выбора данных по нескольким условиям:
либо выбирать данные последовательно, либо сразу по нескольким условиям,
но через оператор &. Отличие оператора and от оператора &: and - выводит
последнее проверенное значение, & - выводит пересечение значений.
Пример: s[ _  &  _ ].sum() 

In [ ]:
np.random.seed(242)
s = pd.Series(np.random.normal(0, 1, 100))
s ** 3
s.index *= 3
s

In [ ]:
s[(s < 2.6) & (s.index % 2 == 1)].sum()

In [ ]:
s[s < 0].shape[0]

Ответы: 1.3801172674551854, 51

### Информация для последующих заданий
- Для всех последующих заданий будем использовать обезличенные транзакционные банковские данные. Для этого считайте в переменные tr_mcc_codes, tr_types, transactions и gender_train из одноимённых таблиц из папки data. Для таблицы transactions используйте только первые n=1000000 строк. Обратите внимание на разделители внутри каждого из файлов - они могут различаться!

In [ ]:
data_transactions = pd.read_csv("data/transactions.csv", sep=',')
data_mcc = pd.read_csv("data/tr_mcc_codes.csv", sep=";")
data_types = pd.read_csv("data/tr_types.csv", sep=";")
data_gender = pd.read_csv("data/gender_train.csv", sep=",")

In [ ]:
data_mcc.head()

In [ ]:
data_types.head()

In [ ]:
data_gender.head()

In [ ]:
data_transactions = data_transactions.head(1000000)
data_transactions.head()

## Описание данных
### Таблица ```transactions.csv```
##### Описание
Таблица содержит историю транзакций клиентов банка за один год и три месяца.

##### Формат данных

```
customer_id,tr_datetime,mcc_code,tr_type,amount,term_id
111111,15 01:40:52,1111,1000,-5224,111111
111112,15 15:18:32,3333,2000,-100,11122233
...
```
##### Описание полей

 - ```customer_id``` — идентификатор клиента;
 - ```tr_datetime``` — день и время совершения транзакции (дни нумеруются с начала данных);
 - ```mcc_code``` — mcc-код транзакции;
 - ```tr_type``` — тип транзакции;
 - ```amount``` — сумма транзакции в условных единицах со знаком; ```+``` — начисление средств клиенту (приходная транзакция), ```-``` — списание средств (расходная транзакция);
 - ```term_id``` — идентификатор терминала;


### Таблица ```gender_train.csv```

##### Описание
Данная таблица содержит информацию по полу для части клиентов, для которых он известен. Для остальных клиентов пол неизвестен.

##### Формат данных
```
customer_id,gender
111111,0
111112,1
...
```

##### Описание полей
 - ```customer_id``` — идентификатор клиента;
 - ```gender``` — пол клиента;

### Таблица ```tr_mcc_codes.csv```

##### Описание
Данная таблица содержит описание mcc-кодов транзакций.

##### Формат данных
```
mcc_code;mcc_description
1000;словесное описание mcc-кода 1000
2000;словесное описание mcc-кода 2000
...
```

##### Описание полей
 - ```mcc_code``` – mcc-код транзакции;
 - ```mcc_description``` — описание mcc-кода транзакции.

### Таблица ```tr_types.csv```

##### Описание
Данная таблица содержит описание типов транзакций.

##### Формат данных
```
tr_type;tr_description
1000;словесное описание типа транзакции 1000
2000;словесное описание типа транзакции 2000
...
```

##### Описание полей
 - ```tr_type``` – тип транзакции;
 - ```tr_description``` — описание типа транзакции;

#### Task 5.
1. В tr_types выберите произвольные 100 строк с помощью метода sample (указав при этом random_seed равный 242)
2. В полученной на предыдущем этапе подвыборке найдите долю наблюдений (стобец tr_description), в которой содержится подстрока 'плата' (в любом регистре). (\*)

Выведите ответ в виде вещественного числа, округлённого до двух знаков после запятой, отделив дробную часть точкой в формате "123.45"

##### Пояснения:
(\*) Строки "ПлатА за аренду", "ПлатАза аренду", "ПЛАТА" удовлетворяют условию, так как будучи переведёнными в нижний регистр содержат подстроку "плата".

In [21]:
np.random.seed(242)
data = data_types.sample(n=100)
data

tr_type                                     tr_description
37      6210                    Возврат покупки.POS Зарубеж. ТУ
50      7034  Перевод на карту/ с карты через АТМ (без взима...
145     2431  Перевод с карты на счет физ.лица в другом банк...
4       4001  Плата за получение наличных в Сбербанке (в дру...
69      8000                   Поправки по счету клиен. Зачисл.
43      7020                           Взнос наличных через POS
5       4010        Плата за получение наличных в АТМ Сбербанка
125     2351                Погашение кредита с карты через ВСП
16      4060  Плата за перевод на карту (с карты) через POS ...
120     2332  Списание с карты по операции “перевода с карты...
147     2433  Перевод с карты на счет физ.лица в другом банк...
64      7082  Перевод/пополнение со счета на карту через Инт...
92      4045                                                н/д
31      5110                            Списание % за овердрафт
48      7031  Перевод на карту (с карты) через АТМ (из одног...
100     1100                                Покупка. ТУ  Россия
78      8150                           Корректировка зачисления
149     2901  Безналичное списание денежных средств со счета...
30      4500                             Плата за превыш.лимита
118     2330  Списание с карты по операции “перевода с карты...
127     2353    Погашение кредита с карты через Сбербанк ОнЛ@йн
110     2100                                  Наличные.  Россия
143     2423  Перевод с карты на счет физ.лица в другом тер....
82      8230  Плата за предоставление по запросу клиента док...
115     2321                           Платеж с карты через ВСП
81      8220  Плата за обработку счета банковской карты клие...
23      4083           Оплата услуг банка по поручению автомат.
139     2420  Перевод с карты на счет физ.лица в другом тер....
72      8003          Списание подоходного налога с нерезидента
7       4020  Плата за получение наличных через POS в Сбербанке
..       ...                                                ...
26      4200              Плата за получение наличных. Зарубеж.
137     2412  Перевод с карты на счет др.лица в одном тер. б...
56      7045  Перевод на карту/ с карты через POS (со взиман...
97      1000                                   Покупка.ТУ СБ РФ
42      7015  Взнос наличных через АТМ (в других ТБ) по счет...
17      4061  Плата за перевод на карту (с карты) через POS ...
129     2370  Списание с карты на карту по операции <перевод...
9       4030  Плата за взнос наличных через АТМ (в своем тер...
94      4600                                                н/д
119     2331  Списание с карты по операции “перевода с карты...
73      8035                             Списание. Счет закрыт.
67      7100               Межфилиальные перечисления процентов
123     2342  Списание с карты по операции “перевода с карты...
58      7060  Пополнение счета для погашения задолженности п...
89      3300                                                н/д
54      7042  Перевод на карту (с карты) через POS (в предел...
131     2400      Перевод средств с карты на счет клиента в ВСП
53      7041  Перевод на карту (с карты) через POS (из одног...
29      4440  Плата за получение через АТМ выписки о 10 посл...
19      4071  Плата за перевод на карту (с карты) через Моби...
85      1510                                                н/д
101     1110                             Покупка. POS ТУ Россия
109     2021  Выдача наличных через POS Сбербанка (в других ТБ)
18      4062  Плата за перевод на карту (с карты) через POS ...
12      4041  Плата за взнос наличных через POS (в другом те...
24      4100                Плата за получение наличных. Россия
130     2371  Списание с карты на карту по операции <перевод...
140     2421  Перевод с карты на счет физ.лица в другом тер....
6       4011  Плата за получение наличных в АТМ Сбербанка (в...
52      7040  Перевод на карту (с карты) через POS (в предел...

[100 rows x 2 columns]

In [22]:
data.tr_description = data.tr_description.apply(lambda x: x.lower())
data_tmp = data.tr_description.apply(lambda x: x.find("плата") != -1)
data_plata = data[data_tmp]
data_plata

tr_type                                     tr_description
4      4001  плата за получение наличных в сбербанке (в дру...
5      4010        плата за получение наличных в атм сбербанка
16     4060  плата за перевод на карту (с карты) через pos ...
30     4500                             плата за превыш.лимита
82     8230  плата за предоставление по запросу клиента док...
81     8220  плата за обработку счета банковской карты клие...
23     4083           оплата услуг банка по поручению автомат.
7      4020  плата за получение наличных через pos в сбербанке
8      4021  плата за получение наличных через pos в сберба...
1      3210  плата за предоставление отчета по счету карты ...
15     4052  плата за перевод на карту (с карты) через атм ...
80     8210                    плата за приост. действия карты
10     4031  плата за взнос наличных через атм (в другом те...
0      3200  плата за предоставление услуг посредством моби...
2      3800  плата за обслуживание банковской карты (за пер...
20     4080                       оплата услуг банка через всп
14     4051  плата за перевод на карту (с карты) через атм ...
26     4200              плата за получение наличных. зарубеж.
17     4061  плата за перевод на карту (с карты) через pos ...
9      4030  плата за взнос наличных через атм (в своем тер...
29     4440  плата за получение через атм выписки о 10 посл...
19     4071  плата за перевод на карту (с карты) через моби...
18     4062  плата за перевод на карту (с карты) через pos ...
12     4041  плата за взнос наличных через pos (в другом те...
24     4100                плата за получение наличных. россия
6      4011  плата за получение наличных в атм сбербанка (в...

In [23]:
frac = data_plata.shape[0] / data.shape[0]
frac

0.26

Ответ: 0.26

#### Task 6. 
1. Для поля tr_type датафрейма transactions посчитайте частоту встречаемости всех типов транзакций tr_type в transactions.
2. Из перечисленных вариантов выберите те, которые попали в топ-5 транзакций по частоте встречаемости.

#### Выберите все верные пункты:
    - 1) Выдача наличных в АТМ Сбербанк России
    - 2) Комиссия за обслуживание ссудного счета
    - 3) Списание по требованию	
    - 4) Оплата услуги. Банкоматы СБ РФ
    - 5) Погашение кредита (в пределах одного филиала)
    - 6) Покупка. POS ТУ СБ РФ

In [24]:
type_counts = data_transactions.tr_type.value_counts()
type_counts = type_counts.to_frame()
type_counts = type_counts.rename(columns={"tr_type" : "count"})
type_counts.index.name = "tr_type"
type_counts

count
tr_type        
1010     231117
2010     151166
7070     149006
1110     137658
1030     118975
2370      49830
7010      28972
7030      26078
7071      15781
1100      15476
1200      10419
2330       9428
2011       6880
4071       6264
2371       6264
1210       5244
4010       4925
7031       3032
4051       2697
2331       2697
7020       2036
4011       1846
2020       1790
6110       1666
2110       1273
4110       1271
7011       1252
2320       1237
2210        680
2440        555
...         ...
4041         68
4031         63
2456         60
2021         57
4500         57
4097         50
4020         36
6210         33
4100         28
7024         26
7040         25
2100         19
6000         17
2340         14
7035         12
7075         12
4021         11
7041          9
7015          8
8145          8
1000          8
2200          6
4200          6
2341          4
2446          4
4061          4
4096          4
8146          3
8100          2
1510          2

[72 rows x 1 columns]

In [25]:
data_types=data_types.set_index("tr_type")
data_types

tr_description
tr_type                                                   
3200     Плата за предоставление услуг посредством моби...
3210     Плата за предоставление отчета по счету карты ...
3800     Плата за обслуживание банковской карты (за пер...
4000               Плата за получение наличных в Сбербанке
4001     Плата за получение наличных в Сбербанке (в дру...
4010           Плата за получение наличных в АТМ Сбербанка
4011     Плата за получение наличных в АТМ Сбербанка (в...
4020     Плата за получение наличных через POS в Сбербанке
4021     Плата за получение наличных через POS в Сберба...
4030     Плата за взнос наличных через АТМ (в своем тер...
4031     Плата за взнос наличных через АТМ (в другом те...
4040     Плата за взнос наличных через POS (в своем тер...
4041     Плата за взнос наличных через POS (в другом те...
4050     Плата за перевод на карту (с карты) через АТМ ...
4051     Плата за перевод на карту (с карты) через АТМ ...
4052     Плата за перевод на карту (с карты) через АТМ ...
4060     Плата за перевод на карту (с карты) через POS ...
4061     Плата за перевод на карту (с карты) через POS ...
4062     Плата за перевод на карту (с карты) через POS ...
4071     Плата за перевод на карту (с карты) через Моби...
4080                          Оплата услуг банка через ВСП
4081                          Оплата услуг банка через АТМ
4082              Оплата услуг банка через Сбербанк ОнЛ@йн
4083              Оплата услуг банка по поручению автомат.
4100                   Плата за получение наличных. Россия
4110             Плата за получение наличных в АТМ. Россия
4200                 Плата за получение наличных. Зарубеж.
4210           Плата за получение наличных в АТМ. Зарубеж.
4400                Плата за получение баланса в чужом АТМ
4440     Плата за получение через АТМ выписки о 10 посл...
...                                                    ...
2351                   Погашение кредита с карты через ВСП
2352                   Погашение кредита с карты через АТМ
2353       Погашение кредита с карты через Сбербанк ОнЛ@йн
2354     Погашение кредита с карты по поручению клиента...
2370     Списание с карты на карту по операции <перевод...
2371     Списание с карты на карту по операции <перевод...
2400         Перевод средств с карты на счет клиента в ВСП
2401     Перевод средств с карты на счет клиента через АТМ
2402     Перевод средств с карты на счет клиента через ...
2403     Перевод средств с карты на счет клиента по пор...
2410     Перевод с карты на счет др.лица в одном тер. б...
2411     Перевод с карты на счет др.лица в одном тер. б...
2412     Перевод с карты на счет др.лица в одном тер. б...
2413     Перевод с карты на счет др.лица в одном тер. б...
2420     Перевод с карты на счет физ.лица в другом тер....
2421     Перевод с карты на счет физ.лица в другом тер....
2422     Перевод с карты на счет физ.лица в другом тер....
999999                                                 XXX
2423     Перевод с карты на счет физ.лица в другом тер....
2430     Перевод с карты на счет физ.лица в другом банк...
2431     Перевод с карты на счет физ.лица в другом банк...
2432     Перевод с карты на счет физ.лица в другом банк...
2433     Перевод с карты на счет физ.лица в другом банк...
2900        Выдача наличных средств без предъявления карты
2901     Безналичное списание денежных средств со счета...
2990                                Списание по требованию
2992     Списание средств для погашения задолженности п...
3001               Комиссия за обслуживание ссудного счета
3030     Плата за обслуживание банковской карты (за пос...
3120                   Плата за досрочный перевыпуск карты

[155 rows x 1 columns]

In [26]:
pd.merge(type_counts, data_types, left_on='tr_type', right_on='tr_type',  how='inner')

count                                     tr_description
tr_type                                                           
1010     231117                              Покупка. POS ТУ СБ РФ
2010     151166              Выдача наличных в АТМ Сбербанк России
7070     149006  Перевод на карту (с карты) через Мобильный бан...
1110     137658                             Покупка. POS ТУ Россия
1030     118975                     Оплата услуги. Банкоматы СБ РФ
2370      49830  Списание с карты на карту по операции <перевод...
7010      28972       Взнос наличных через АТМ (в своем тер.банке)
7030      26078  Перевод на карту (с карты) через АТМ (в предел...
7071      15781  Перевод на карту (с карты) через Мобильный бан...
1100      15476                                Покупка. ТУ  Россия
1200      10419                               Покупка. Зарубеж. ТУ
2330       9428  Списание с карты по операции “перевода с карты...
2011       6880      Выдача наличных в АТМ Сбербанка (в других ТБ)
4071       6264  Плата за перевод на карту (с карты) через Моби...
2371       6264  Списание с карты на карту по операции <перевод...
1210       5244                          Покупка. POS  Зарубеж. ТУ
4010       4925        Плата за получение наличных в АТМ Сбербанка
7031       3032  Перевод на карту (с карты) через АТМ (из одног...
4051       2697  Плата за перевод на карту (с карты) через АТМ ...
2331       2697  Списание с карты по операции “перевода с карты...
7020       2036                           Взнос наличных через POS
4011       1846  Плата за получение наличных в АТМ Сбербанка (в...
2020       1790                Выдача наличных через POS Сбербанка
6110       1666                     Возврат покупки. POS ТУ Россия
2110       1273                             Наличные.   АТМ Россия
4110       1271          Плата за получение наличных в АТМ. Россия
7011       1252                           Взнос наличных через АТМ
2320       1237     Безналичный перевод денежных средств через POS
2210        680                       Наличные.  АТМ Зарубеж. банк
2440        555                                                н/д
...         ...                                                ...
1410        179                                                н/д
1310        166                                                н/д
4090        144                                                н/д
7034        126  Перевод на карту/ с карты через АТМ (без взима...
4041         68  Плата за взнос наличных через POS (в другом те...
4031         63  Плата за взнос наличных через АТМ (в другом те...
2021         57  Выдача наличных через POS Сбербанка (в других ТБ)
4500         57                             Плата за превыш.лимита
4020         36  Плата за получение наличных через POS в Сбербанке
6210         33                    Возврат покупки.POS Зарубеж. ТУ
4100         28                Плата за получение наличных. Россия
7024         26  Взнос наличных через POS (в своем ТБ) по счету...
7040         25  Перевод на карту (с карты) через POS (в предел...
2100         19                                  Наличные.  Россия
6000         17                         Возврат  покупки. ТУ СБ РФ
2340         14  Списание с карты по операции “перевода с карты...
7035         12  Перевод на карту/ с карты через АТМ (со взиман...
7075         12  Перевод с карты на карту в овердрафте через Мо...
4021         11  Плата за получение наличных через POS в Сберба...
7041          9  Перевод на карту (с карты) через POS (из одног...
7015          8  Взнос наличных через АТМ (в других ТБ) по счет...
8145          8      Зачисление до завершения претензионной работы
1000          8                                   Покупка.ТУ СБ РФ
2200          6                           Наличные.  Зарубеж. банк
4200          6              Плата за получение наличных. Зарубеж.
2341          4  Списание с карты по операции “перевода с карты...
4061          4  Плата за перевод на карту (с карты) чере

Ответ: 1, 4, 6

#### Task 7.
1. В датафрейме transactions задайте столбец  customer_id в качестве индекса.
2. Выделите клиента с максимальной суммой транзакции (то есть с максимальным приходом на карту). (\*)
3. Найдите у него наиболее часто встречающийся модуль суммы приходов/расходов. (\**)


#### Выберите все верные пункты:
    - 1) 1122957.89
    - 2) 15721.41
    - 3) 22459.16
    - 4) 13475494.63
    - 5) 107407.78
    - 6) 65019.26

##### Пояснения:
(\*) Если у клиента были транзакции [-10000, 10, 0, -10], то максимумом будет являться значение 10. <br>
(\**) Если у клиента были транзакции [-10000, 10, 0, -10], то наиболее встречающийся модуль суммы транзакций равен 10, и встретился он 2 раза.

In [27]:
from scipy.stats import mode
data_trans=data_transactions.set_index("customer_id")
data_trans_am = data_trans[['amount']]
gb = data_trans_am.groupby(['customer_id'])
gb = gb['amount'].unique()
gb = gb.to_frame()
gb['max'] = gb['amount'].apply(lambda x: x.max())
gb['mod_mode'] = gb['amount'].apply(lambda x: mode(np.abs(x))[0][0])
gb.sort_values(by='max', ascending=False).head()

amount           max  \
customer_id                                                                    
26782922     [-31442.82, -11935.25, -11229.58, -67377.47, -...  1.459845e+08   
19950741     [-19921.27, -108949.37, -168443.68, -161537.49...  1.347549e+08   
40917212     [-3795.6, -28087.42, -17383.39, -33688.74, -14...  6.737747e+07   
82581973     [-8961.2, -11229.58, -3503.63, -5495.53, -6092...  6.737747e+07   
77695202     [-42223.22, -4522943.99, -1583032.83, 1572141....  6.123489e+07   

             mod_mode  
customer_id            
26782922      2245.92  
19950741     22459.16  
40917212       118.58  
82581973      5614.79  
77695202     67377.47

In [28]:
data_trans
data_trans.sort_values(by='amount', ascending=False).head()

tr_datetime  mcc_code  tr_type        amount term_id
customer_id                                                       
26782922     151 11:35:44      6010     7070  1.459845e+08     NaN
19950741     143 14:08:24      6010     7070  1.347549e+08     NaN
82581973     192 18:18:17      6010     7020  6.737747e+07     NaN
40917212      43 17:10:40      6010     7070  6.737747e+07     NaN
77695202     132 14:07:18      6010     7070  6.123489e+07     NaN

Максимальный приход 1.459845e+08 с модой модуля 2245.92
Второй по значению приход 1.347549e+08 с модой модуля 22459.16
Ответ: 3, 4 (Наиболее подходящие для 2 в топе значения)

#### Task 8.
1. Найдите максимальную разницу между медианами суммы транзакций, посчитанными при заданных ниже условиях по полю amount из таблицы transactions (\*):
    - Медиана суммы транзакций
    - Медиана суммы транзакций по тем строкам, которые ни в одном из своих столбцов не содержат пустые значения
    - Медиана суммы транзакций по строкам, отсортированным по полю amount в порядке возрастания, и из которых удалены дублирующиеся по столбцам [mcc_code, tr_type] строки, причём при удалении соответстующих дублей остаются только последние из дублирующихся строк (keep='last')

Выведите ответ в виде вещественного числа, округлённого до двух знаков после запятой, отделив дробную часть точкой в формате "123.45"

##### Пояснения:
(\*) Для вычисления максимальной разницы между значениями списка можно использовать функцию np.ptp  
(\**) Если в результате получились значения [1,3,5], то макимальная разница между ними 4 == 5-1.

In [30]:
from statistics import median
data_trans=data_transactions.set_index("customer_id")
data_trans_am = data_trans[['amount']]
gb = data_trans_am.groupby(['customer_id'])
gb = gb['amount'].unique()
gb = gb.to_frame()
gb['median'] = gb['amount'].apply(lambda x: median(x))
np.ptp(gb['median'])

731045.58

In [31]:
data_trans=data_transactions.set_index("customer_id")
data_trans=data_trans.dropna()
data_trans_am = data_trans[['amount']]
gb = data_trans_am.groupby(['customer_id'])
gb = gb['amount'].unique()
gb = gb.to_frame()
gb['median'] = gb['amount'].apply(lambda x: median(x))
np.ptp(gb['median'])

1009539.135

In [32]:
data_trans=data_transactions.set_index("customer_id")
data_trans=data_trans.drop_duplicates(subset=['mcc_code', 'tr_type'], keep='last')
data_trans=data_trans.sort_values(by='amount', ascending=True)
data_trans

tr_datetime  mcc_code  tr_type      amount    term_id
customer_id                                                        
85273426     375 00:00:00      8062     1110 -3833778.22   77047223
14264900     146 12:15:12      4829     2340 -3368873.66        NaN
30627668     318 00:00:00      8299     1200 -3011268.84   00000002
99717689     379 14:03:48      4722     1010 -2078595.05   10306231
36837532     258 00:00:00      5970     1210 -2048585.79   CC100201
20795048     364 00:00:00      6010     2200 -1451029.13   80772600
77695202     160 00:00:00      8398     1210 -1269301.31        NaN
17389182     153 00:00:00      7999     1200 -1264136.15        NaN
63145287     253 10:47:30      6010     2020 -1235253.67     089038
14892960     104 00:00:00      4722     1200  -954205.61        NaN
18654396     426 12:15:51      6300     1210  -948073.37   PVTE3225
26782922     246 23:30:14      3501     1200  -898404.04   00424211
2740197      433 15:44:00      5681     1010  -898366.31   10270069
56436494     250 14:49:48      5712     1100  -869349.08   I0MO0IRQ
17389182     126 00:00:00      3501     1100  -834806.89        NaN
99717689     169 00:00:00      5611     1110  -762600.70        NaN
67901119     168 00:00:00      8011     1210  -698106.76        NaN
93358299     257 00:00:00      6211     1100  -673774.73   20001041
17923253     401 10:29:01      5733     1110  -660074.65   17037025
1585299      433 08:00:27      5532     1100  -655807.41     759823
63145287     314 09:47:09      4829     2331  -645543.57      80140
8942066      454 00:00:00      5722     1110  -604151.34   21749902
54645432     167 00:00:00      5722     1200  -597189.00        NaN
17389182     158 00:00:00      5611     1210  -584641.07        NaN
54890147      91 00:00:00      5261     1100  -570328.75        NaN
99717689     202 00:00:00      7011     1200  -558413.94        NaN
48111232     200 00:00:00      4511     1210  -545907.34        NaN
14892960      21 00:00:00      5944     1200  -543511.62        NaN
91582559      55 00:00:00      3351     1210  -533989.83        NaN
49933875     116 00:00:00      5044     1200  -510396.04        NaN
...                   ...       ...      ...         ...        ...
95202077     312 00:00:00      5691     6100   127006.54   00000000
2368868      264 00:00:00      5722     6100   134530.35   00000000
63145287     454 13:02:14      6010     7031   134754.95     633710
2658384       57 00:00:00      5661     6110   149128.81        NaN
62030642     333 18:30:31      5699     6100   166130.39        NaN
87629740     215 17:09:50      5039     6110   166646.95   50121459
77695202     384 00:00:00      5691     6210   174058.47   50125058
30627668     383 00:00:00      7999     6200   176975.24        NaN
56639375     148 20:12:43      5655     6010   179560.97        NaN
81998653     398 12:03:25      7299     6200   182724.34        NaN
63145287     348 00:00:00      6012     6110   186411.01   00020156
14892960      90 00:00:00      4722     6200   196998.71        NaN
89619970     105 23:38:49      3000     6000   204153.74        NaN
85821996     216 00:00:00      4814     8100   221357.46  001-CADJT
20930910      57 00:00:00      5511     6110   224591.58        NaN
17331961     191 19:23:11      5661     6010   224591.58        NaN
49856670      17 00:00:00      5722     6110   226612.90        NaN
57546321      27 00:00:00      5211     6110   229328.44        NaN
17389182      54 00:00:00      4511     6110   250419.61        NaN
34267056     176 00:00:00      6011     8145   320145.86        NaN
36460428     257 19:14:26      5681     6010   322131.70   10070096
57765004     145 14:10:05      5611     6010   343400.52        NaN
73342030     431 00:00:00      7512     6200   380604.79   04PSCVOI
37633014     205 16:02:49      6011     7011   561478.94        NaN
29268985     280 07:23:04      8099     6010   576751.17     333687
37633014     448 13:30:42      6010     7021   622567.85     878293
17

In [34]:
data_trans_am = data_trans[['amount']]
gb = data_trans_am.groupby(['customer_id'])
gb = gb['amount'].unique()
gb = gb.to_frame()
gb['median'] = gb['amount'].apply(lambda x: median(x))
np.ptp(gb['median'])

3945624.83

Ответ: 
1) 731045.58 
2) 1009539.135
3) 3945624.83